File: src/main/java/de/zalando/zally/rules/SnakeCaseInPropNameRule.java
package de.zalando.zally.rules;

import de.zalando.zally.Violation;
import de.zalando.zally.ViolationType;
import io.swagger.models.Model;
import io.swagger.models.Swagger;
import io.swagger.models.properties.Property;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class SnakeCaseInPropNameRule implements Rule {

    //only accepts lower case and underscores
    private String pattern = "^[a-z][a-z_]*$";
    String title = "snake_case property names";
    String description = "Property names must be snake_case (and never camelCase)";
    String ruleLink = "http://zalando.github.io/restful-api-guidelines/naming/Naming.html" +
            "#must-use-snakecase-never-camelcase-for-query-parameters";

    public List<Violation> validate(Swagger swagger) {

        List<Violation> violations = new ArrayList<Violation>();

        if (swagger.getDefinitions() == null) {
            return violations;
        }

        Map<String, Model> definitions = swagger.getDefinitions();
        for (String definitionName : definitions.keySet()) {
            Map<String, Property> properties = definitions.get(definitionName).getProperties();
            for (String propertyName : properties.keySet()) {
                if (!isSnakeCase(propertyName)) {
                    violations.add(new Violation(title, description + "\n Violating property: " + propertyName + " in definition: " + definitionName, ViolationType.MUST, ruleLink));
                }
            }
        }

        return violations;
    }

    boolean isSnakeCase(String input) {
        return input.matches(pattern);
    }
}

File: server/src/main/java/de/zalando/zally/rule/SnakeCaseInPropNameRule.kt
package de.zalando.zally.rule

import com.typesafe.config.Config
import de.zalando.zally.dto.ViolationType
import de.zalando.zally.util.PatternUtil
import de.zalando.zally.util.getAllJsonObjects
import io.swagger.models.Swagger
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

@Component
class SnakeCaseInPropNameRule(@Autowired rulesConfig: Config) : SwaggerRule() {
    override val title = "snake_case property names"
    override val url = "http://zalando.github.io/restful-api-guidelines/json-guidelines/JsonGuidelines.html" +
        "#must-property-names-must-be-snakecase-and-never-camelcase"
    override val violationType = ViolationType.MUST
    override val code = "M012"
    private val description = "Property names must be snake_case: "

    private val whitelist = rulesConfig.getStringList(SnakeCaseInPropNameRule::class.simpleName + ".whitelist").toSet()

    override fun validate(swagger: Swagger): Violation? {
        val result = swagger.getAllJsonObjects().flatMap { (def, path) ->
            val badProps = def.keys.filterNot { PatternUtil.isSnakeCase(it) || whitelist.contains(it) }
            if (badProps.isNotEmpty()) listOf(badProps to path) else emptyList()
        }
        return if (result.isNotEmpty()) {
            val (props, paths) = result.unzip()
            val properties = props.flatten().toSet().joinToString(", ")
            Violation(this, title, description + properties, violationType, url, paths)
        } else null
    }
}
